### PR TITLE
feat: updates from uproot discussion

### PR DESCRIPTION
**Pull request by henryiii**

### CODE REVIEW

**Comment by henryiii** [comment]
Refinements from a long discussion with @jpivarski today. @HDembinski, let me know what you think - otherwise, this is what is going into Uproot 4's first release candidate next week, and boost-histogram soon.

Key changes:

* Dropped Optional arrays - now all three methods are always return an array; instead of None, you can query `weighted` and `interpretation` if you are a smart plot function, and "dumb" plot functions that just plot values and variances do the right thing, and don't need to check `is None` everywhere.
* Tightened the types just a bit - an individual axis will always return the same thing from `__getitem__`.

We went with `weighted` instead of `is_weighed` though it was just about a tie. This seemed a little more similar to the rest of the names. We didn't add a `count_variances`, thought a future revision could have something like that too.

File: uhi/typing/plottable.py [hunk 1]
```diff
@@ -46,10 +59,20 @@ def __eq__(self, other: Any) -> bool:
...
+PlottableAxisStr = PlottableAxisGeneric[str]
+
+PlottableAxis = Union[PlottableAxisContinuous, PlottableAxisInt, PlottableAxisStr]
+
+
 @runtime_checkable
 class PlottableHistogram(Protocol):
     axes: Sequence[PlottableAxis]
 
+    weighted: bool
```
Comment (HDembinski) [inline uhi/typing/plottable.py:73]: I don't see why `weighted` has to be exposed. Whether the histogram is weighted or not should be an implementation detail, the plotter simply requests values and variances. It does not need to know whether the content is weighted.

File: uhi/typing/plottable.py [hunk 2]
```diff
@@ -46,10 +59,20 @@ def __eq__(self, other: Any) -> bool:
...
+    interpretation: Literal["count", "mean", "median"]
```
Comment (HDembinski) [inline uhi/typing/plottable.py:74]: Better use an enum?

File: uhi/typing/plottable.py [hunk 3]
```diff
@@ -58,15 +81,16 @@ def values(self) -> ArrayLike:
         All methods can have a flow=False argument - not part of Protocol.
         """
 
-    def variances(self) -> Optional[ArrayLike]:
+    def variances(self) -> ArrayLike:
         """
-        Variance returns the variance if applicable, otherwise None.
+        Variance returns the variance. If unweighed, this is identical to .counts().
```
Comment (HDembinski) [inline uhi/typing/plottable.py:86]: 👍

File: uhi/typing/plottable.py [hunk 4]
```diff
@@ -58,15 +81,16 @@ def values(self) -> ArrayLike:
...
         """
 
-    def counts(self) -> Optional[ArrayLike]:
+    def counts(self) -> ArrayLike:
         """
-        Count returns the number of values in a mean accumulator (also known as
-        a Profile histogram), or is None for normal storages.
+        Count returns the number of values counted in a mean accumulator (also
+        known as a Profile histogram), or is identical to .values if
+        interpretation is "count".
```
Comment (HDembinski) [inline uhi/typing/plottable.py:95]: 👍

**Review by HDembinski — COMMENTED** [review]
This looks very good to me. I would only only change the `interpretation` from string to enum. And drop the `weighted` property.

File: uhi/typing/plottable.py (hunk 2 above)
Comment (HDembinski) [reply uhi/typing/plottable.py:74]: Do we have a histogram that computes the median already?

Comment (henryiii) [reply uhi/typing/plottable.py:74]: This can't be an enum because it's a protocol; where would you put it? If you are plotting an Uproot histogram with mplhep, you can't depend an an enum from boost-histogram, for example. And "median" isn't supported by boost-histogram (yet?), but Jim wanted it for future proofing since it's a common type that is used in some other areas.

File: uhi/typing/plottable.py (hunk 1 above)
Comment (henryiii) [reply uhi/typing/plottable.py:73]: You need something that can inform a smart plotter that "variances" are useful new information or just values again. A dumb plotter can just always plot variances, but a smarter one should be able to avoid drawing error bars if you are not using a weighted storage. I would want a plotter to add error bars if I use a Weight() storage, but not if I use a Double() storage, it would be plotting incorrect information if I filled with a weight. Maybe the name is not ideal? `has_variance` or something like that?

Basically, this should not plot error bars by default, as they are wrong:

```python
h = Hist(..., storage=Double())
h.fill(values, weight=weights)
mplhep.histplot(h)
```

Changing `Double()` to `Weight()` should then add the error bars, or maybe explicitly forcing them.

Comment (henryiii) [reply uhi/typing/plottable.py:73]: That's why this was Optional, but we moved to just providing some way for a smart plotting function to tell if this is really new information or not.

Comment (HDembinski) [reply uhi/typing/plottable.py:67]: You always want error bars, even for normal histograms where variance = value.

Comment (HDembinski) [reply uhi/typing/plottable.py:67]: `has_variance` makes more sense then, but it is still weird to have variance() always return an array. If that is optional, it should be None if there is no variance information and an array otherwise. I still think that `has_variance` should not be there. Python has a natural way to return optional results (return the result or None).

File: uhi/typing/plottable.py (hunk 2 above)
Comment (HDembinski) [reply uhi/typing/plottable.py]: @jpivarski Where is median a common type?

Comment (HDembinski) [reply uhi/typing/plottable.py]: I think the protocol right now should only contain the least common denominator of what is needed now. More literals can be added to `interpretation` at any later time.

Comment (HDembinski) [reply uhi/typing/plottable.py]: > This can't be an enum because it's a protocol; where would you put it?

The protocol requires interface to be present, it can also require that an enum with certain field exists.

Comment (HDembinski) [reply uhi/typing/plottable.py]: Every library would have their own enum, just like every library has their own classes that implement the protocol. Strings are bad.

Comment (henryiii) [reply uhi/typing/plottable.py]: Python uses string comparisons everywhere, it's harder to find examples of enums than of strings. For example, matplotlib's `plt.plot` has strings everywhere, such as`'default', 'steps', 'steps-pre', 'steps-mid', 'steps-post'` for drawstyle, etc. So using a string would be a reasonable Pythonic thing to do.

> it can also require that an enum with certain field exists.

No, it cannot. For example:

```python
class Enum1(enum.Enum):
    one = 1
class Enum2(enum.Enum):
    one = 1

Enum1.one == Enum2.one
# FALSE  - these are unique
```

The point of an enum is that is unique, you can't make one up without using it. Since the whole point of a Protocol is to talk between libraries without introducing a dependence, this won't work! The plotting library's enum would not match the source library's enum.

If you are comparing the string-name of the enum fields, besides misusing enums in a way they were not designed, that's also much less convenient than just using a string in the first place!

Also, while performance doesn't matter - this is not even a problem either. In Python, short strings are interned, and when you compare, it just compares the pointer. Both of these are short enough:

```python
>>> id("mean")
4482494000
>>> id("mean")
4482494000
>>> id("count")
4479637616
>>> id("count")
4479637616
```

Now, when defining this in a library, I _think_ a valid implementation would be:

```python
class StorageType(str, enum.Enum):
    count = "count"
    mean = "mean"
```

But I _think_ this is a valid implementation of this, and would be up to the library implementing it. (Need to verify it would pass static and dynamic isinstance correctly, but I think it should - might be a bug if it doesn't.)

Comment (henryiii) [reply uhi/typing/plottable.py]: PS: If it's not clear from context, a str mixin enum compares with the value, that is, `StorageType.count == "count"` is True.

File: uhi/typing/plottable.py (hunk 1 above)
Comment (henryiii) [reply uhi/typing/plottable.py:67]: Maybe the most important thing was fixing `.counts` to always return, and we could let `.variances` be Optional. @jpivarski?

File: uhi/typing/plottable.py (hunk 2 above)
Comment (henryiii) [reply uhi/typing/plottable.py]: Ugh, ran into an issue if this is implemented as a string enum by a client library. `Enum.count` is not valid for a [StrEnum](https://docs.python.org/3/library/enum.html#others) - as `.count` is defined on a string, so Enum.count.count doesn't make any sense - the first count should be an enum instance and the second count would be a string method, but it's actually Enum.count again. How about `sum` and `mean` instead? Or all-caps?

Comment (jpivarski) [reply uhi/typing/plottable.py]: The reason I was against "sum" is because an accumulator might be written that sums all `x` values given by `fill(x)` or `y` in `fill(x, y)` (times any weights given as an optional last parameter, not shown here). That's not what this does: it sums `1` (possibly times weight) for every `fill(x)`, with the `x` value only used to determine which bin it should go in. So Histogrammar distinguished between (possibly weighted) "count" and "sum", though "count" is by far more common.

One option would be to pluralize all of the choices: "counts," maybe "sums," definitely "means" (for profiles). Or maybe use more traditional language. Only "counts" gives you a true histogram, so "histogram" vs "profile" vs other names. (What, in that scheme, would the name for medians be?)

Or maybe capitalization if the equality of these strings is taken to be case-insensitive, with the capitalized form as canonical. Then the enums would appear in capitalized form (not unexpected, since they're constant symbols), but strings are allowed to be lowercase (which feels more natural—or maybe just more modern).

File: uhi/typing/plottable.py [hunk 5]
```diff
@@ -1,35 +1,71 @@
...
 
-Plotters should see if .counts() is None - if not, this is a Profile
-Then check .variances; if not None, this storage holds variance information and
-error bars should be included. If it is None, then a plotter can avoid showing
-error bars (recommended), or use np.sqrt(h.values()).
+Producers: use isinstance(myhist, PlottableHistogram) in your tests; part of
+the protocol is checkable at runtime, though ideally you should use MyPy; if
+your histogram class supports PlottableHistogram, this will pass.
+
+Consumers: Make your functions the accept PlottableHistogram static type, and
```
Comment (henryiii) [inline uhi/typing/plottable.py:8]:
```suggestion
Consumers: Make your functions accept the PlottableHistogram static type, and
```

**Comment by henryiii** [comment]
Let's merge, then open a PR for the final, possible change.
